cpp/ql/src/semmle/code/cpp/dataflow/internal/tainttracking1/TaintTrackingImpl.qll

```diff
@@ -42,57 +35,79 @@ private import TaintTrackingParameter::Private
+   */
   // overridden to provide taint-tracking specific qldoc
   abstract override predicate isSource(DataFlow::Node source);
 
-  /** Holds if `sink` is a taint sink. */
+  /**
+   * Holds if `sink` is a relevant taint sink.
+   *
+   * The smaller this predicate is, the faster `hasFlow()` will converge.
+   */
   // overridden to provide taint-tracking specific qldoc
   abstract override predicate isSink(DataFlow::Node sink);
 
-  /**
-   * Holds if taint should not flow into `node`.
-   */
+  /** Holds if the node `node` is a taint sanitizer. */
   predicate isSanitizer(DataFlow::Node node) { none() }
 
-  /** Holds if data flow from `node1` to `node2` is prohibited. */
-  predicate isSanitizerEdge(DataFlow::Node node1, DataFlow::Node node2) {
-    none()
+  final override predicate isBarrier(DataFlow::Node node) {
+    isSanitizer(node) or
+    defaultTaintBarrier(node)
+  }
+
+  /** DEPRECATED: override `isSanitizerIn` and `isSanitizerOut` instead. */
+  deprecated predicate isSanitizerEdge(DataFlow::Node node1, DataFlow::Node node2) { none() }
+
+  deprecated final override predicate isBarrierEdge(DataFlow::Node node1, DataFlow::Node node2) {
+    isSanitizerEdge(node1, node2)
   }
 
+  /** Holds if data flow into `node` is prohibited. */
+  predicate isSanitizerIn(DataFlow::Node node) { none() }
+
+  final override predicate isBarrierIn(DataFlow::Node node) { isSanitizerIn(node) }
+
+  /** Holds if data flow out of `node` is prohibited. */
+  predicate isSanitizerOut(DataFlow::Node node) { none() }
+
+  final override predicate isBarrierOut(DataFlow::Node node) { isSanitizerOut(node) }
+
+  /** Holds if data flow through nodes guarded by `guard` is prohibited. */
+  predicate isSanitizerGuard(DataFlow::BarrierGuard guard) { none() }
+
+  final override predicate isBarrierGuard(DataFlow::BarrierGuard guard) { isSanitizerGuard(guard) }
+
   /**
-   * Holds if the additional taint propagation step
-   * from `source` to `target` must be taken into account in the analysis.
-   * This step will only be followed if `target` is not in the `isSanitizer`
-   * predicate.
+   * Holds if the additional taint propagation step from `node1` to `node2`
+   * must be taken into account in the analysis.
    */
-  predicate isAdditionalTaintStep(DataFlow::Node source,
-                                  DataFlow::Node target)
-  { none() }
-
-  final override
-  predicate isBarrier(DataFlow::Node node) { isSanitizer(node) }
+  predicate isAdditionalTaintStep(DataFlow::Node node1, DataFlow::Node node2) { none() }
 
-  /** DEPRECATED: use `isSanitizerEdge` instead. */
-  override deprecated
-  predicate isBarrierEdge(DataFlow::Node node1, DataFlow::Node node2) {
-    this.isSanitizerEdge(node1, node2)
+  final override predicate isAdditionalFlowStep(DataFlow::Node node1, DataFlow::Node node2) {
+    isAdditionalTaintStep(node1, node2) or
+    localAdditionalTaintStep(node1, node2) or
+    globalAdditionalTaintStep(node1, node2)
```

Comment:
Could we replace `localAdditionalTaintStep(node1, node2) or globalAdditionalTaintStep(node1, node2)` with `defaultAdditionalTaintStep(node1, node2)` in the shared code?

Reply:
That's fine with me. What do you think, @hvitved ?

Reply:
Sounds fine with me.

Reply:
Done.